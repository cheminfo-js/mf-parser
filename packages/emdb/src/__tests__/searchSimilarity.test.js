'use strict';

const DBManager = require('..');

describe('test searchSimilarity', () => {
  it('should find one result with bad distribution', () => {
    let dbManager = new DBManager();
    dbManager.loadNeutralTest({ maxC: 10 });
    dbManager.setExperimentalSpectrum({ x: [41, 121], y: [1, 1] });
    let results = dbManager.searchSimilarity({
      ionizations: 'H+,(H+)2,(H+)3', // useless because the test database has already ionizations
      filter: {},
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 2.5,
        },
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });

    expect(results.test).toHaveLength(2);

    expect(results.test).toMatchSnapshot();
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.886, 3);
    expect(results.test[1].ms.similarity.value).toBeCloseTo(0.886, 3);
  });

  it('should find one result with perfect match (small zone)', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {},
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 0.5,
        },
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    //   expect(results.test).toMatchSnapshot();
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with bad distribution, small zone, small width', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 0.5,
        },
        widthBottom: 0.1,
        widthTop: 0.1,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });

    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find no result because of filter', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1,
      },
    });
    expect(results.test).toHaveLength(0);
  });

  it('should find one result with bad bad distribution, large window', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 2.5,
        },
        widthBottom: 0.1,
        widthTop: 0.1,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.895, 2);
  });

  it('should find one result with bad bad distribution, large window huge width ', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        zone: {
          low: -0.5,
          high: 2.5,
        },
        widthBottom: 5,
        widthTop: 5,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with overlap', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120], y: [1] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        zone: {
          low: -0.5,
          high: 2.5,
        },
        common: 'first', // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBe(1);
  });

  it('should find one result with good distribution', () => {
    let dbManager = new DBManager();
    dbManager.loadTest();
    dbManager.setExperimentalSpectrum({ x: [120, 121], y: [1, 0.11] });
    let results = dbManager.searchSimilarity({
      ionizations: '+',
      filter: {
        msem: 120,
        precision: 1000,
      },
      similarity: {
        widthBottom: 0.05,
        widthTop: 0.01,
        common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
      },
    });
    expect(results.test[0].ms.similarity.value).toBeCloseTo(0.995, 2);
  });
});
