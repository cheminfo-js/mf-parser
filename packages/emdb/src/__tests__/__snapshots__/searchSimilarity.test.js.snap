// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test searchSimilarity should find one result with bad distribution 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 10,
    },
    "charge": 0,
    "em": 120,
    "ionization": Object {
      "charge": 3,
      "em": 3.02347509669,
      "mf": "(H+)3",
    },
    "mf": "C10",
    "ms": Object {
      "charge": 3,
      "em": 41.00727645232093,
      "ionization": "(H+)3",
      "similarity": Object {
        "difference": Array [
          Array [
            41.00727645232093,
            41.34172806401093,
            41.67617967570093,
          ],
          Array [
            0.011955179514731795,
            0.09713957777327428,
            0.004727858758499098,
          ],
        ],
        "experimental": Array [
          Array [
            41,
          ],
          Array [
            1,
          ],
        ],
        "experimentalInfo": Object {
          "from": 40.840609785654266,
          "max": 0.5,
          "min": 0.5,
          "sum": 0.5,
          "to": 41.840609785654266,
        },
        "quantity": 0.5000694876998228,
        "theoretical": Array [
          Array [
            41.00727645232093,
            41.34172806401093,
            41.67617967570093,
          ],
          Array [
            0.8981325634682266,
            0.09713957777327428,
            0.004727858758499098,
          ],
        ],
        "thereoticalInfo": Object {
          "max": 0.8980077624805509,
          "min": 0.004727201793740607,
          "sum": 0.9998610439118322,
        },
        "value": 0.8861773839534948,
      },
    },
    "mw": 120.10735896735248,
    "parts": Array [
      "C10",
    ],
    "unsaturation": 11,
  },
  Object {
    "atoms": Object {
      "C": 10,
    },
    "charge": 0,
    "em": 120,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C10",
    "ms": Object {
      "charge": 1,
      "em": 121.00727645232094,
      "ionization": "H+",
      "similarity": Object {
        "difference": Array [
          Array [
            121.00727645232094,
            122.01063128739094,
            123.01398612246093,
          ],
          Array [
            0.011955179515087067,
            0.09713957777327428,
            0.004727858758499098,
          ],
        ],
        "experimental": Array [
          Array [
            121,
          ],
          Array [
            1,
          ],
        ],
        "experimentalInfo": Object {
          "from": 120.50727645232094,
          "max": 0.5,
          "min": 0.5,
          "sum": 0.5,
          "to": 123.50727645232094,
        },
        "quantity": 0.5000694876998228,
        "theoretical": Array [
          Array [
            121.00727645232094,
            122.01063128739094,
            123.01398612246093,
          ],
          Array [
            0.8981325634682266,
            0.09713957777327428,
            0.004727858758499098,
          ],
        ],
        "thereoticalInfo": Object {
          "max": 0.8980077624805509,
          "min": 0.004727201793740607,
          "sum": 0.9998610439118322,
        },
        "value": 0.8861773839531395,
      },
    },
    "mw": 120.10735896735248,
    "parts": Array [
      "C10",
    ],
    "unsaturation": 11,
  },
]
`;
